### ISSUE TITLE
Tests in dynamic feature module fail when run through gradle command line

### ISSUE THREAD

**Issue by RafaO** [android/app-bundle-samples#11]
I wrote some tests for a Dynamic Feature Module inside `DynamicFeature/src/test/java/my/package/MyTest.kt`. These tests rely on a class that is hosted in the main `app` module. If I run the tests using the Android Studio configuration they run without problems, but on the other hand if I use `./gradlew test` they fail saying that the class on the main module can't be found: 
```
    java.lang.NoClassDefFoundError
        Caused by: java.lang.ClassNotFoundException
```

In order to include the tests inside my CI, I would need either to know how to get the command line order that Android Studio is using, or figure out why when running the gradle task, it is not able to find the main module app.

Any help appreciated.
Thanks in advance!

**Comment by wangchauyan** [comment]
Anyone has the solution for this issue?

**Comment by luisansal** [comment]
I have the same issue, no solution yet.
Any solution? please?

**Comment by wangchauyan** [comment]
@luisansal  I currently do such things: 
1. make sure the DFMs have the same Gradle setting as the App module. 
2. make sure the DFMs can access the classes of the App Module. 

Then it works when I run the Gradle CLI

Another thought is: 
Maybe we can create interfaces in the App Module and implement these interfaces in DFMs, but that depends on how your project structures.

**Comment by luisansal** [comment]
Hi Mr Wang,

App Module classes, not found from other dynamic feature module.
I did the same gradle settings as the App module, but that doesn't work for me.

Team another solution?.

**Comment by luisansal** [comment]
@wangchauyan Can you please, share a part of your gradle code config, or ./gradlew command line?

**Comment by wangchauyan** [comment]
@luisansal  
Hi,

Sorry, I can't share the code snap for you because it's internal codebase.
Here are my understanding of your situation and possible solutions, 
I think your situation is that you have some classes in the App module. 
And your DFM (let's say `auth`) use these classes from the App module.
You can write the test cases in `auth` module and import these classes from the App module. 

But, when running the test case in `auth` module, you might sometimes encounter a few issues. 

1. Can't resolve the supertype of the class in the App module 
-> This means you might need to have the same Gradle implementation in your `auth` module cause the App module depends on some libraries. 
2. NoClassDefFoundError
-> This means you might need to separate these shared classes out from the App module to a shared module for example, and you might need to create a jar file for your `auth` module or you can try something like this : 

```
testImplementation project(path: ':App', configuration: 'tests')
```

in your `auth` module. 
Hope this helps.

**Comment by keyboardsurfer** [comment]
The code under test needs to be accessible from the dynamic feature module's (DFM) test classes.
To enable this it either has to be in the DFM or included using `androidTest(Implementation|Api)`. A way to keep it reusable throughout and not hard wire it to the DFM is to create a `com.android.library` module and depend on that.

**Comment by masterjefferson** [comment]
It would be nice to get a complete explanation of why this issue occurs. Android Studio can run the DFM tests without issue, and it seems to be manually passing the classpaths as CLI arguments. While refactoring code into a separate library module might fix it, in some cases this is a non-trivial task. We should not need to refactor significant amounts of our codebase just to be able to duplicate IDE functionality in a CI environment.

cc @jeffchang5

**Comment by keyboardsurfer** [comment]
@masterjefferson Can you please let me know which versions of Android Studio & build tools you're using for this?

**Comment by masterjefferson** [comment]
@keyboardsurfer 
I have tried both AS 3.5.3 and 3.6.0 RC 1 (with respective gradle plugin versions)

Build tools is 29.0.2, but also occurs with 28.0.3.

Gradle wrapper is using version 5.6.4.

**Comment by aednlaxer** [comment]
Same issue for me. Feature module tests fail with `NoClassDefFoundError` when running them in terminal on local machine or in CI. Tests pass in Android Studio.

I'm using AS 3.5.3, build tools 29.0.2, Gradle 5.6.4

**Comment by keyboardsurfer** [comment]
We're currently in the progress of testing a fix for this issue.

**Comment by keyboardsurfer** [comment]
We're currently in the progress of testing a fix for this issue.

**Comment by keyboardsurfer** [comment]
This issue has been fixed and will be rolled out in one of the next Android Studio 4.0 canary releases.

**Comment by chrisjenx** [comment]
@keyboardsurfer is this hot-patchable to 3.6? we are having to disable tests in feature modules because of this.

**Comment by keyboardsurfer** [comment]
Hi @chrisjenx! Unfortunately this can't be easily patched as there were major changes. Thus the version code change from 3.x to 4.x.  
Your best call is to upgrade to 4.x once you're ready for it or have two versions run in parallel.

**Comment by MagsMagnoli** [comment]
@keyboardsurfer I just updated my project to Android Studio 4 and `4.0.0-beta01` gradle plugin and am still seeing `NoClassDefFoundError`s related to `FragmentDirections` - the generated `NavDirections` for my fragment which is located in a dynamic feature module

**Comment by keyboardsurfer** [comment]
It should be okay from beta02. I am on 4.1 Canary 2 and things work well here.

**Comment by MagsMagnoli** [comment]
Updated to beta02 but I'm on 4.0beta for AS and still seeing issues
I resolve it by making a dummy class in the dynamic feature module
Will check on AS 4.1 C2

**Comment by ulx** [comment]
@MagsMagnoli 
I think it helps you
 // This "test" source set is a fix for SafeArgs classes not being available when running Unit tests from cmd
    // See: https://issuetracker.google.com/issues/139242292
    sourceSets {
        getByName("test").java.srcDir("${project.rootDir}/app/build/generated/source/navigation-args/debug")
    }

**Comment by keyboardsurfer** [comment]
There was another regression in the canaries. It has been resolved and should be shipped with Canary 6.

**Comment by thapelo96** [comment]
i am currently dealing with this problem, nothing is working out

**Comment by keyboardsurfer** [comment]
@thapelo96 Can you confirm that the issue is reproducible when running `./gradlew connectedAndroidTest --no-parallel` using the [sample code](https://github.com/android/app-bundle-samples/tree/master/DynamicFeatures)?

**Comment by soygabimoreno** [comment]
I have the same issue on AS 4.0 even using this configuration:
`testRuntimeOnly(files("$projectDir$"/../app/build/intermediates/app_classes/debug/classes.jar"))`

I have tried AS 4.1 Canary 10 and I have checked it works.

Is there any way to make it work on AS 4.0?

**Comment by soygabimoreno** [comment]
> I have the same issue on AS 4.0 even using this configuration:
> `testRuntimeOnly(files("$projectDir$"/../app/build/intermediates/app_classes/debug/classes.jar"))`
> 
> I have tried AS 4.1 Canary 10 and I have checked it works.
> 
> Is there any way to make it work on AS 4.0?

My fault! I have missed to add another dynamic feature with the `testRuntimeOnly` way. So, I have something like this:
`testRuntimeOnly(files("$projectDir$"/../app/build/intermediates/app_classes/debug/classes.jar"))`
`testRuntimeOnly(files("$projectDir$"/../<dynamicFeatureName>/build/intermediates/app_classes/debug/classes.jar"))`

**Comment by gaurigadkari** [comment]
I am still seeing this error `Caused by: java.lang.ClassNotFoundException:` on Canary 4.2

**Comment by soygabimoreno** [comment]
> I am still seeing this error `Caused by: java.lang.ClassNotFoundException:` on AS 4.2 Canary 4

It has to be resolved at Canary 6.

**Comment by mochadwi** [comment]
> > I have the same issue on AS 4.0 even using this configuration:
> > `testRuntimeOnly(files("$projectDir$"/../app/build/intermediates/app_classes/debug/classes.jar"))`
> > I have tried AS 4.1 Canary 10 and I have checked it works.
> > Is there any way to make it work on AS 4.0?
> 
> My fault! I have missed to add another dynamic feature with the `testRuntimeOnly` way. So, I have something like this:
> `testRuntimeOnly(files("$projectDir$"/../app/build/intermediates/app_classes/debug/classes.jar"))`
> `testRuntimeOnly(files("$projectDir$"/../<dynamicFeatureName>/build/intermediates/app_classes/debug/classes.jar"))`

For those still want to use `4.0` to supports product flavours 

Modify the code into 

```
testRuntimeOnly(files("$projectDir$"/../<baseApp>/build/intermediates/app_classes/<buildVariant>/classes.jar"))
testRuntimeOnly(files("$projectDir$"/../<dynamicFeatureName>/build/intermediates/app_classes/<buildVariant>/classes.jar")
```

Also make sure for `/../` to have a correct path. 

E.g with below directory structures:

```
|baseApp
|features
|--------/dynamicFeatureName

// needs to `cd` change directory twice to parent
testRuntimeOnly(files("$projectDir$"/../../<baseApp>/build/intermediates/app_classes/<buildVariant>/classes.jar"))
testRuntimeOnly(files("$projectDir$"/../../<dynamicFeatureName>/build/intermediates/app_classes/<buildVariant>/classes.jar")
```

**Comment by AmritZoman** [comment]
Not working for me even in Android Studio 4.2 Canary 7.. 

Also tried on AS 4.0.1 by adding the below line, but no luck
testRuntimeOnly(files("$projectDir/../app/build/intermediates/app_classes/internalDebug/classes.jar"))
testRuntimeOnly(files("$projectDir/../<dynamicFeatureName>/build/intermediates/app_classes/internalDebug/classes.jar"))

The error is same: Unresolved reference: BaseTest (Trying to refer BaseTest present in app inside the dfm's tests)

Please help here

**Comment by gaurigadkari** [comment]
I tried the Android Studio 4.2 Canary 7 as well, still doesn't work.

**Comment by gaurigadkari** [comment]
> > I am still seeing this error `Caused by: java.lang.ClassNotFoundException:` on AS 4.2 Canary 4
> 
> It has to be resolved at Canary 6.

I tried the Android Studio 4.2 Canary 7 as well, still doesn't work.

**Comment by chrisjenx** [comment]
> Not working for me even in Android Studio 4.2 Canary 7..
> 
> Also tried on AS 4.0.1 by adding the below line, but no luck
> testRuntimeOnly(files("$projectDir/../app/build/intermediates/app_classes/internalDebug/classes.jar"))
> testRuntimeOnly(files("$projectDir/..//build/intermediates/app_classes/internalDebug/classes.jar"))
> 
> The error is same: Unresolved reference: BaseTest (Trying to refer BaseTest present in app inside the dfm's tests)
> 
> Please help here

You can't refer to test classes in another modules tests - that's not related to this issue. (I'm guessing your BaseTest is in `app/src/tests` and your accessing it via `dfm/src/tests` ?)

**Comment by chrisjenx** [comment]
@keyboardsurfer is there a new issue for this? As it shouldn't really be closed?

**Comment by chrisjenx** [comment]
FYI for all of those struggling with this workaround:

For library classes the path is different

```
testRuntimeOnly(files("${projectDir}/../../<library module path>/build/intermediates/compile_library_classes_jar/< buildVariant >/classes.jar"))
```

**Comment by keyboardsurfer** [comment]
I do not see this issue affecting the sample at hand.  
Executing both `./gradlew test` or `./gradlew connectedAndroidTest` from within the `DynamicFeatures` project does not yield any failed tests.  
In order to see what the actual issue is, I would have to see it either being reproducible in one of the samples in this repository or with a [MCVE](https://stackoverflow.com/help/minimal-reproducible-example).

**Comment by JabezNzomo99** [comment]
Adding the line below to the dynamic feature module gradle file solved the issue on AS 4.1.1 (STABLE) Anyone knows whether the issue has been solved on any AS canary builds?

`testRuntimeOnly(files("$projectDir/../app/build/intermediates/app_classes/staging/classes.jar"))`

**Comment by mochadwi** [comment]
:alhamdulillah: upgrading to AGP 7.2, running test in DFM works without the above workaround somehow. (no `testRuntimeOnly`
